core/base/src/TSystem.cxx

```diff
@@ -1144,11 +1144,11 @@ Bool_t TSystem::ExpandFileName(const char *fname, char *xname, const int kBufSiz
          ++c;
       }
    } else if (c[0] == '~' && c[1] != '/') { // ~user case
-      buff[0] = 0;
       n = strcspn(c+1, "/ ");
-      if ((size_t) n >= sizeof(buff))
-         n = sizeof(buff) - 1;
-      (void) strlcat(buff, std::string(c+1).c_str(), n);
+      buff[0] = 0;
+      assert(n < kBufSize && "This should have been prevented by the truncation 'strlcat(inp, c, kBufSize)'");
+      // There is no overlap here as the buffer is segment in 4 strings of at most kBufSize
+      strlcat(buff, c+1, n);
```

Comment:
Please keep the cast to `(void)` or we will eventually start getting warnings about unused return value of `strlcat()`.

Reply:
Another comment is why don't we just use the [`wordexp()`](http://man7.org/linux/man-pages/man3/wordexp.3.html) function?

Reply:
wordexp: maybe ... one would have to check whether it has the same behavior and if it is available on Windows ....

void: okay

>  I actually think that my original implementation with sizeof(buff) was correct

The 'point' of this line is to extra the username from the string '~username/pathname' by definition we *must* cut it off at the 'n'-th character ...

Reply:
Then we should probably go back to using `strncpy()`.

Reply:
yes, here strlcpy and strlcat are equivalent.  I will be using strlcpy.
strncpy is not an option anymore (even-though it is the 'right' choice) because of the weak gcc8 warning detection engine.

Reply:
The problem was not the warning detection. It was a real bug to use the length of the source, since you can overflow the buffer where things are being copied. But I agree that using `strlcpy()` should work.

Reply:
> The problem was not the warning detection. 

It is ...  The following code still warns:
```
strncpy(output, input, std::min( n, sizeof(output));
```
I am not found any way to use strncpy to do string chopping without warning in gcc8.